### PR TITLE
[Wip] Enforce "exp" and "nbf" claim format

### LINKED ISSUE #17: Enforcing a stricter kind of validation
Would you be against a stricter kind of jwt validation?

Currently when the `exp`|`nbf` claim are not expressed as a number, they're silently ignored.

Two options:

 - Make it stricter by design and fail the verification of the jwt with a meaningful message when exp isn't a number
 - Allow an optional stricter kind of validation (through `jwt:enforce_strict_validation(bool)` for instance)

### PR DESCRIPTION
**Pull request by fermaem**
Fix #17

This can be tuned down through the `loose_validation(bool)` configuration function

### CODE REVIEW

File: lib/resty/jwt.lua [hunk 1]
```diff
@@ -452,28 +466,51 @@ end
...
   local exp = jwt_obj[str_const.payload][str_const.exp]
   local nbf = jwt_obj[str_const.payload][str_const.nbf]
 
-  if (exp ~= nil or nbf ~= nil ) and not jwt_obj[str_const.reason] then
-    leeway = leeway or 0
-    local now = ngx.now()
+  leeway = leeway or 0
+  local now = ngx.now()
+
+  if exp ~= nil and not is_validation_loose then
```
Comment (SkyLothar) [inline lib/resty/jwt.lua]: if `is_validation_loose` is `true`.
`exp` checking is disabled.

but what we really want to do

present | valid | loose | result
------- | ------- | ------- | -------
√ | √ | - | perform checking
√ | x | √ | skip the checking
√ | x | x | raise 500 with meaningful message
x | - | - | skip the checking

is that right :question:

Comment (nulltoken) [inline lib/resty/jwt.lua]: @SkyLothar I'm actually not sure about the **raise 500** result.

Rationale is that this I don't see this as a server issue. The jwt is malformed and we requested a strict validation, thus, we should only reject the jwt.

IOW, I'm not sure to see the point of a 500 as it's there's nothing that can actually be done server side to "fix" the jwt.

Given this, we somehow end with the following behaviors (which I think is that the proposed code implements).

| present | loose validation | result |
|  :--: |  :---: | :--: |
| :white_check_mark: | :heavy_multiplication_x: | perform checking, fail the verification if malformed |
| :white_check_mark: | :white_check_mark:  | skip the checking |
| :heavy_multiplication_x:  | :heavy_multiplication_x:  | skip the checking |
| :heavy_multiplication_x:  | :white_check_mark:   | skip the checking |

It's also completely possible that I misunderstood your point or that there's an obvious flaw in my reasoning (That wouldn't be the first time :wink:)

Thoughts?

**Comment by nulltoken** [comment]
Another angle on this PR would be: Do we need a loose kind of validation? What would actually lead someone to set up a jwt validation mechanism and decide to turn the loose validation mechanism on?

The way the .Net version of this library implements this is to not allow a loose kind of validation but rather, to describe what should be validated (exp/nbf, issuer, ...).

See **[this Apache 2.0 licenced piece of code](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/System.IdentityModel.Tokens.Jwt/TokenValidationParameters.cs#L104-L133)** for more details.

**Comment by fermaem** [comment]
Another angle on this PR would be: Do we need a loose kind of validation? What would actually lead someone to set up a jwt validation mechanism and decide to turn the loose validation mechanism on?

The way the .Net version of this library implements this is to not allow a loose kind of validation but rather, to describe what should be validated (exp/nbf, issuer, ...).

See **[this Apache 2.0 licenced piece of code](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/System.IdentityModel.Tokens.Jwt/TokenValidationParameters.cs#L104-L133)** for more details.

File: lib/resty/jwt.lua (hunk 1 above)
Comment (fermaem) [inline lib/resty/jwt.lua]: @SkyLothar I'm actually not sure about the **raise 500** result.

Rationale is that this I don't see this as a server issue. The jwt is malformed and we requested a strict validation, thus, we should only reject the jwt.

IOW, I'm not sure to see the point of a 500 as it's there's nothing that can actually be done server side to "fix" the jwt.

Given this, we somehow end with the following behaviors (which I think is that the proposed code implements).

| present | loose validation | result |
|  :--: |  :---: | :--: |
| :white_check_mark: | :heavy_multiplication_x: | perform checking, fail the verification if malformed |
| :white_check_mark: | :white_check_mark:  | skip the checking |
| :heavy_multiplication_x:  | :heavy_multiplication_x:  | skip the checking |
| :heavy_multiplication_x:  | :white_check_mark:   | skip the checking |

It's also completely possible that I misunderstood your point or that there's an obvious flaw in my reasoning (That wouldn't be the first time :wink:)

Thoughts?

**Comment by SkyLothar** [comment]
@fermaem @nulltoken 
I think `loose` is more like a `smart validation`.
All proper formed claims will be validated automatically.

But, as you pointed out, `loose validation` makes no sense.
A malformed JWT is an invalid JWT, it should be rejected.

.Net's idea looks good. I should implement it too! :+1:

**Comment by SkyLothar** [comment]
So, yeah, this PR is good for now.

Separated validators for each claim is coming soon.

**Comment by fermaem** [comment]
> Separated validators for each claim is coming soon.

:cool: :bangbang:

**Pull request merged**